File: python/university_security/models/pedagog.py
"""Pedagog domain class — mirrors Pedagog.java."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime

from .actor import Actor


@dataclass
class Pedagog(Actor):
    """An academic staff member / lecturer.

    Attributes
    ----------
    pedagog_id:        Unique identifier (e.g. ``"P001"``).
    emri:              First name.
    mbiemri:           Last name.
    email:             University e-mail address.
    fjalekalimi:       Hashed / plain password (demo only).
    departamenti:      Academic department.
    titulliAkademik:   Academic title — ``"Prof."``, ``"Dr."``, ``"Asoc."`` …
    gjendja_llogarise: Account state — ``"Aktive"``, ``"Locked"``,
                       or ``"UnderReview"``.
    date_regjistrimi:  Registration timestamp (default: *now*).
    """

    pedagog_id: str
    emri: str
    mbiemri: str
    email: str
    fjalekalimi: str
    departamenti: str
    titulliAkademik: str
    gjendja_llogarise: str
    date_regjistrimi: datetime = field(default_factory=datetime.now)

    # ── Actor abstract property ───────────────────────────────────────────────
    @property
    def roli(self) -> str:
        return "Pedagog"

    # ── Auth behaviours ───────────────────────────────────────────────────────
    def login(self, email: str, password: str) -> bool:
        return self.email == email and self.fjalekalimi == password

    def logout(self) -> None:
        print(f"Pedagog {self.emri} doli.")

    def verifiko_2fa(self, kod: str) -> bool:
        return kod is not None and len(kod) == 6

    # ── Domain behaviours ─────────────────────────────────────────────────────
    def fut_note(self, nota: str) -> None:
        print(f"Nota u fut: {nota}")

    def ngarko_material(self, material: str) -> None:
        print(f"Material ngarkuar: {material}")

    # ── Dunder methods ────────────────────────────────────────────────────────
    def __str__(self) -> str:
        return (
            f"[{self.pedagog_id}] {self.titulliAkademik} {self.emri}"
            f" {self.mbiemri} — {self.departamenti}"
        )

    def __repr__(self) -> str:
        return (
            f"Pedagog(pedagog_id={self.pedagog_id!r}, emri={self.emri!r},"
            f" mbiemri={self.mbiemri!r}, email={self.email!r},"
            f" departamenti={self.departamenti!r},"
            f" titulliAkademik={self.titulliAkademik!r},"
            f" gjendja_llogarise={self.gjendja_llogarise!r})"
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Pedagog):
            return NotImplemented
        return self.pedagog_id == other.pedagog_id

    def __hash__(self) -> int:
        return hash(self.pedagog_id)

File: Pedagog.java
import java.util.Date;

public class Pedagog {
    private String pedagogId;
    private String emri;
    private String mbiemri;
    private String email;
    private String fjalëkalimi;
    private String roli = "Pedagog";
    private String departamenti;
    private String titulliAkademik;  // Prof., Dr., Asoc.
    private String gjendjaLlogarise;
    private Date   datëRegjistrimi;

    public Pedagog() {}

    public Pedagog(String pedagogId, String emri, String mbiemri,
                   String email, String fjalëkalimi, String departamenti,
                   String titulliAkademik, String gjendjaLlogarise) {
        this.pedagogId        = pedagogId;
        this.emri             = emri;
        this.mbiemri          = mbiemri;
        this.email            = email;
        this.fjalëkalimi      = fjalëkalimi;
        this.departamenti     = departamenti;
        this.titulliAkademik  = titulliAkademik;
        this.gjendjaLlogarise = gjendjaLlogarise;
        this.datëRegjistrimi  = new Date();
    }

    public boolean login(String email, String pass) {
        return this.email.equals(email) && this.fjalëkalimi.equals(pass);
    }

    public void logout()                    { System.out.println("Pedagog " + emri + " doli."); }
    public boolean verifiko2FA(String kod)  { return kod != null && kod.length() == 6; }
    public void futNote(String nota)        { System.out.println("Nota u fut: " + nota); }
    public void ngarkoMaterial(String mat)  { System.out.println("Material ngarkuar: " + mat); }

    // ── Getters & Setters ──────────────────────────────────────────────────
    public String getPedagogId()           { return pedagogId; }
    public void   setPedagogId(String v)    { this.pedagogId = v; }
    public String getEmri()                { return emri; }
    public void   setEmri(String v)         { this.emri = v; }
    public String getMbiemri()             { return mbiemri; }
    public void   setMbiemri(String v)      { this.mbiemri = v; }
    public String getEmail()               { return email; }
    public void   setEmail(String v)        { this.email = v; }
    public String getFjalëkalimi()         { return fjalëkalimi; }
    public void   setFjalëkalimi(String v)  { this.fjalëkalimi = v; }
    public String getRoli()                { return roli; }
    public String getDepartamenti()        { return departamenti; }
    public void   setDepartamenti(String v) { this.departamenti = v; }
    public String getTitulliAkademik()     { return titulliAkademik; }
    public void   setTitulliAkademik(String v){ this.titulliAkademik = v; }
    public String getGjendjaLlogarise()    { return gjendjaLlogarise; }
    public void   setGjendjaLlogarise(String v){ this.gjendjaLlogarise = v; }
    public Date   getDatëRegjistrimi()     { return datëRegjistrimi; }

    @Override
    public String toString() {
        return "[" + pedagogId + "] " + titulliAkademik + " " + emri + " " + mbiemri + " — " + departamenti;
    }
}
